### ISSUE TITLE
Code blocks in documentation search

### ISSUE THREAD

**Issue by tbrlpld** [wagtail/wagtail#10139]
When using search on the docs, the surfaced results seem to ignore words that are found in code blocks (inline or block). 

E.g. the search for `image_url` only reveals two entries: 

https://docs.wagtail.org/en/stable/search.html?q=image_url

<img width="300" alt="image" src="https://user-images.githubusercontent.com/24797493/220660753-1d1f5b59-c653-490a-883e-6da5e8052d17.png">

The second result contains the search term more accurately and multiple times. The other instances are not shown in the results view. Not sure if this is because of the page was already linked. But in that case it would be nice to see the more fitting results first.

https://docs.wagtail.org/en/stable/advanced_topics/images/image_serve_view.html#generating-dynamic-image-urls-in-python

<img width="300" alt="Screenshot 2023-02-22 at 06 59 49" src="https://user-images.githubusercontent.com/24797493/220661173-537d305e-54be-49e3-9a61-4f73256ef786.png">

But, there are other pages in the docs that the search completely missed.

https://docs.wagtail.org/en/stable/advanced_topics/performance.html#image-urls

<img width="300" alt="image" src="https://user-images.githubusercontent.com/24797493/220661526-5bc20560-e018-4cbd-89ff-e8c14b2d9d4d.png">

I wonder if we configure Algolia somehow to pick up code blocks better. 

----

```[tasklist]
### Tasks
- [ ] #10207
- [ ] #8159
- [x] Devise 10-20 queries to compare the performance of different search engines
- [ ] Set up testing protocol to compare Algolia DocSearch with and without code block indexing, and RTD search
- [ ] Share results as a spreadsheet for review
```

**Comment by gasman** [comment]
This is a known problem with the Algolia search indexing along with #8159, although it seems that we didn't have an open issue for it already, so thanks for reporting!

**Comment by thibaudcolas** [comment]
Thanks for the report @tbrlpld! Can confirm this is the case, though only for code blocks. Inline code formatting is still searchable as long as it’s inside a heading, paragraph, or list item.

For code blocks – this is intentional as per recommendations from Algolia, who state indexing code blocks creates a lot of noise since there’s lots of repetition in code.

We’ve discussed this at the last core team meeting and decided to give this a go anyway, as we now have much more control over how the indexing of the docs is configured. So we’ll be able to compare results with and without code blocks indexed.

**Comment by tbrlpld** [comment]
Awesome. Thanks for the update @thibaudcolas

**Comment by thibaudcolas** [comment]
Now ready to be picked up (was waiting on #8159). If anyone has suggested searches to try this out with please post them here.

While investigating this I think I might have also spotted another related issue: `dt` elements which we use for Python classes don’t seem to be indexed either.

**Comment by thibaudcolas** [comment]
Here is the comparison, with results from Algolia as-is, Algolia with code blocks indexed, and Read the Docs – side by side across 44 different queries: https://wagtail-docs-search-comparison.netlify.app/.

The code blocks indexing gives code blocks relatively low priority over keywords appearing in headings, so the difference isn’t too big, but still there. Searches with differences:

- https://wagtail-docs-search-comparison.netlify.app/#404
- https://wagtail-docs-search-comparison.netlify.app/#chosefield
- https://wagtail-docs-search-comparison.netlify.app/#embedblock
- https://wagtail-docs-search-comparison.netlify.app/#emoji
- https://wagtail-docs-search-comparison.netlify.app/#parentalkey
- https://wagtail-docs-search-comparison.netlify.app/#use_json_field
- https://wagtail-docs-search-comparison.netlify.app/#ChoiceBlock
- https://wagtail-docs-search-comparison.netlify.app/#ImageChooserBlock
- https://wagtail-docs-search-comparison.netlify.app/#modeladmin_register
- https://wagtail-docs-search-comparison.netlify.app/#ClusterableModel
- https://wagtail-docs-search-comparison.netlify.app/#get_children

Based on this I’d say I generally find results with code blocks better than those without. The difference is rarely big (aside from [get_children](https://wagtail-docs-search-comparison.netlify.app/#get_children)), but code blocks seem to have the advantage of adding extra results when there might not be anything otherwise.

**Comment by tbrlpld** [comment]
Interesting, it still does not seem to find this page: Dynamic image serve view — Wagtail Documentation 4.2.1 documentationdocs.wagtail.orgwhen looking for “image_url”On Mar 20, 2023, at 15:11, Thibaud Colas ***@***.***> wrote:﻿
Here is the comparison, with results from Algolia as-is, Algolia with code blocks indexed, and Read the Docs – side by side across 44 different queries: https://wagtail-docs-search-comparison.netlify.app/.
The code blocks indexing gives code blocks relatively low priority over keywords appearing in headings, so the difference isn’t too big, but still there. Searches with differences:

https://wagtail-docs-search-comparison.netlify.app/#404
https://wagtail-docs-search-comparison.netlify.app/#chosefield
https://wagtail-docs-search-comparison.netlify.app/#embedblock
https://wagtail-docs-search-comparison.netlify.app/#emoji
https://wagtail-docs-search-comparison.netlify.app/#parentalkey
https://wagtail-docs-search-comparison.netlify.app/#use_json_field
https://wagtail-docs-search-comparison.netlify.app/#ChoiceBlock
https://wagtail-docs-search-comparison.netlify.app/#ImageChooserBlock
https://wagtail-docs-search-comparison.netlify.app/#modeladmin_register
https://wagtail-docs-search-comparison.netlify.app/#ClusterableModel
https://wagtail-docs-search-comparison.netlify.app/#get_children

Based on this I’d say I generally find results with code blocks better than those without. The difference is rarely big (aside from get_children), but code blocks seem to have the advantage of adding extra results when there might not be anything otherwise.

**Comment by thibaudcolas** [comment]
@tbrlpld I’ve added this specific example at https://wagtail-docs-search-comparison.netlify.app/#image_url

I think that result is there already, even in the first screenshot you shared in this issue? We get one more result with code search turned on but it seems to be a partial match on "image" rather than a match of `image_url`.

**Comment by tbrlpld** [comment]
Right yea, the page is there. I guess I am confused because it's listed because of the sub string match on "generate_image_url" instead of the exact match of "image_url" which is also multiple times on the page. 

But right, it looks like it now highlights one exact match too 👍

**Comment by allcaps** [comment]
Searching for `base_form_class`, I'd expect to be in the results:
- [Customising the editing interface](https://docs.wagtail.org/en/latest/advanced_topics/customisation/page_editing_interface.html)
- [Using forms in admin views](https://docs.wagtail.org/en/latest/advanced_topics/customisation/page_editing_interface.html)

Currently, all Algolia results point to the Panel API.

---

As a comparison it might sometimes be useful to Google search with: `site:docs.wagtail.org/en/latest/ SEARCH_QUERY`

**Comment by thibaudcolas** [comment]
Thanks Coen, I’ve added this specific query to the comparison. It’s a very interesting one because it highlights how Google, Algolia, RTD differ in what they index / how they return results:

- Google gives you one result per relevant page
- Algolia gives you one result per relevant section of a page
- RTD gives you one result per relevant page, but has relevant sections within that parent result

There are a couple things we could try in Algolia to improve its results:

- Only index signatures’ class names / function names, not the parameters – so [Panel API](https://docs.wagtail.org/en/stable/reference/panel_api.html#wagtail.admin.panels.Panel) results are no longer there
- Index class attributes as content rather than sections of the page – so there is a single [Panel API](https://docs.wagtail.org/en/stable/reference/panel_api.html#wagtail.admin.panels.Panel) result.
- Avoid indexing those heading levels separately (`indexHeadings` config option) – unsure of the result

**Comment by thibaudcolas** [comment]
Just a heads’up we’re now opted into [Read the Docs Addons](https://about.readthedocs.com/blog/2024/04/enable-beta-addons/), which has yet another search index available within its "flyover" menu. It’s a "live results" typeahead interface rather than autocomplete, so a big plus over previous Read the Docs implementation, but unclear about the index.

For now we’ve tried:

- `image_url` – prefer the Algolia results as they’re better presented
- `base_form_class` – prefer the RTD addons results as the first result is more relevant

**Comment by gasman** [comment]
Closing as per #12521, as we've ditched Algolia search.
